### ISSUE TITLE
VNC-Session suddenly closes

### ISSUE THREAD

**Issue by masi-d** [TigerVNC/tigervnc#1362]
**Describe the bug**
VNC-Session suddenly closes

**To Reproduce**
Not easy to reproduce, but seems to happen always after a mouse click is done.
Sometimes it happens short after a session started. Sometimes the session is open several minutes or hours before it happens.
And mostly it does not happen at all after hours of usage.

**Expected behavior**
Session does not close suddenly

**Client:**
 - OS: Windows 10 [1909]
 - VNC client: TigerVNC Viewer 64-bit
 - VNC client version: v1.11.0 Built on: 2020-09-08 23:48
 - Client downloaded from: TigerVNC-github

**Server:**
 - OS: Centos 8 (last updated a few weeks ago)
 - VNC server: TigerVNC
 - VNC server version: 1.11.0-6.el8
 - Server downloaded from: Centos Repository
 - Server was started using xinetd:
service vncserver
{
disable = no
socket_type = stream
protocol = tcp
group = tty
wait = no
user = nobody
server = /usr/bin/Xvnc
server_args = -inetd -query localhost -geometry 1280x1024 -depth 24 -once -Log *:syslog:30 -securitytypes=none  -desktop [servername]
}

**Workaround**
I downgraded the VNC-Server to Tiger-VNC 1.10.1 as the problem arised after a systemupdate.
Now the problem seems to be gone. Client remained on 1.11.0.

I tried Server 1.11.0 and Client 1.10.1 to find out where the bug is and here the problem occured also.
So the bug seems to be in the server component.

**Server-LOG (Time: UTC, LOG-level: 30) /var/log/messages:**
```
Oct 28 09:32:30 [servername] Xvnc[2097314]: vncext: VNC extension running!
Oct 28 09:32:30 [servername] Xvnc[2097314]: vncext: created VNC server for screen 0
Oct 28 09:32:30 [servername] Xvnc[2097314]: Connections: accepted: Y.Y.Y.Y::62673
Oct 28 09:32:30 [servername] Xvnc[2097314]: vncext: added inetd sock
Oct 28 09:32:30 [servername] Xvnc[2097314]: SConnection: Client needs protocol version 3.8
Oct 28 09:32:30 [servername] Xvnc[2097314]: SConnection: Client requests security type None(1)
Oct 28 09:32:30 [servername] Xvnc[2097314]: VNCSConnST: Server default pixel format depth 24 (32bpp) little-endian rgb888
Oct 28 09:32:31 [servername] Xvnc[2097314]: ComparingUpdateTracker: 0 pixels in / 0 pixels out
Oct 28 09:32:31 [servername] Xvnc[2097314]: ComparingUpdateTracker: (1:-nan ratio)
Oct 28 09:32:33 [servername] Xvnc[2097314]: ComparingUpdateTracker: 5.24288 Mpixels in / 1.33168 Mpixels out
Oct 28 09:32:33 [servername] Xvnc[2097314]: ComparingUpdateTracker: (1:3.93704 ratio)
Oct 28 09:37:18 [servername] Xvnc[2097314]: VNCSConnST: closing Y.Y.Y.Y::62673: read: Connection reset by peer (104)
Oct 28 09:37:18 [servername] Xvnc[2097314]: EncodeManager: Framebuffer updates: 1749
Oct 28 09:37:18 [servername] Xvnc[2097314]: EncodeManager:   Tight:
Oct 28 09:37:18 [servername] Xvnc[2097314]: EncodeManager:     Solid: 1.013 krects, 20.6435 Mpixels
Oct 28 09:37:18 [servername] Xvnc[2097314]: EncodeManager:            15.8281 KiB (1:5095.38 ratio)
Oct 28 09:37:18 [servername] Xvnc[2097314]: EncodeManager:     Bitmap RLE: 487 rects, 548.89 kpixels
Oct 28 09:37:18 [servername] Xvnc[2097314]: EncodeManager:                 18.0459 KiB (1:119.13 ratio)
Oct 28 09:37:18 [servername] Xvnc[2097314]: EncodeManager:     Indexed RLE: 3.294 krects, 17.8231 Mpixels
Oct 28 09:37:18 [servername] Xvnc[2097314]: EncodeManager:                  1.87456 MiB (1:36.2898 ratio)
Oct 28 09:37:18 [servername] Xvnc[2097314]: EncodeManager:   Tight (JPEG):
Oct 28 09:37:18 [servername] Xvnc[2097314]: EncodeManager:     Full Colour: 390 rects, 4.66846 Mpixels
Oct 28 09:37:18 [servername] Xvnc[2097314]: EncodeManager:                  1.38248 MiB (1:12.885 ratio)
Oct 28 09:37:18 [servername] Xvnc[2097314]: EncodeManager:   Total: 5.184 krects, 43.6839 Mpixels
Oct 28 09:37:18 [servername] Xvnc[2097314]: EncodeManager:          3.29012 MiB (1:50.6669 ratio)
Oct 28 09:37:18 [servername] Xvnc[2097314]: TcpSocket: unable to get peer name for socket
Oct 28 09:37:18 [servername] Xvnc[2097314]: Connections: closed: ::0
Oct 28 09:37:18 [servername] Xvnc[2097314]: ComparingUpdateTracker: 185.875 Mpixels in / 39.2692 Mpixels out
Oct 28 09:37:18 [servername] Xvnc[2097314]: ComparingUpdateTracker: (1:4.73334 ratio)
Oct 28 09:37:18 [servername] Xvnc[2097314]: ComparingUpdateTracker: 0 pixels in / 0 pixels out
Oct 28 09:37:18 [servername] Xvnc[2097314]: ComparingUpdateTracker: (1:-nan ratio)
```

**Client-LOG (Time: GMT+2, LOG-level: 100):**
```
Thu Oct 28 11:32:28 2021
 Win32TouchHandler: The 'Tablet PC Input' service is required for touch
 Win32TouchHandler: Two touch points required, system currently supports: 0

Thu Oct 28 11:32:30 2021
 DecodeManager: Detected 8 CPU core(s)
 DecodeManager: Creating 4 decoder thread(s)
 TcpSocket:   Connecting to X.X.X.X [X.X.X.X] port 5900
 CConn:       Connected to host X.X.X.X port 5900
 CConnection: reading protocol version
 CConnection: Server supports RFB protocol version 3.8
 CConnection: Using RFB protocol version 3.8
 CConnection: processing security types message
 CConnection: Server offers security type None(1)
 CConnection: Choosing security type None(1)
 CConnection: processing security message
 CConnection: processing security result message
 CConnection: processing security result message
 CConnection: Authentication success!

Thu Oct 28 11:32:31 2021
 CConnection: reading server initialisation
 CConnection: initialisation done
 CConn:       Using pixel format depth 24 (32bpp) little-endian rgb888
 Win32TouchHandler: The 'Tablet PC Input' service is required for touch
 Win32TouchHandler: Two touch points required, system currently supports: 0
 CConnection: Enabling continuous updates

Thu Oct 28 11:32:33 2021
 DesktopWindow: Requesting framebuffer resize from 1280x1024 to 1793x1057
 DesktopWindow: 1 screen(s)
 DesktopWindow:     1804289383 (0x6b8b4567): 1793x1057+0+0 (flags 0x00000000)
 DesktopWindow: 
 Viewport:    Resizing framebuffer from 1280x1024 to 1793x1057

Thu Oct 28 11:32:34 2021
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ

Thu Oct 28 11:32:35 2021
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Unexpected release of key code 30
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ

Thu Oct 28 11:32:36 2021
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ

Thu Oct 28 11:32:37 2021
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ

Thu Oct 28 11:32:39 2021
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ

Thu Oct 28 11:34:20 2021
 CConn:       Throughput 13277 kbit/s - changing to quality 6
 Config:      set QualityLevel(Int) to 6

Thu Oct 28 11:34:24 2021
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ

Thu Oct 28 11:34:27 2021
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ

Thu Oct 28 11:34:29 2021
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ

Thu Oct 28 11:34:30 2021
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ

Thu Oct 28 11:37:01 2021
 Viewport:    Sending clipboard data (9 bytes)
 Viewport:    Sending clipboard data (9 bytes)

Thu Oct 28 11:37:02 2021
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ

Thu Oct 28 11:37:04 2021
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ

Thu Oct 28 11:37:05 2021
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ

Thu Oct 28 11:37:06 2021
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ

Thu Oct 28 11:37:07 2021
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ

Thu Oct 28 11:37:08 2021
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ

Thu Oct 28 11:37:09 2021
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ

Thu Oct 28 11:37:10 2021
 Viewport:    Key pressed: 0xYYYY => 0xZZZZ
 Viewport:    Key released: 0xYYYY => 0xZZZZ

Thu Oct 28 11:37:15 2021
 Viewport:    Sending clipboard data (7 bytes)
 Viewport:    Sending clipboard data (7 bytes)

Thu Oct 28 11:37:16 2021
 Viewport:    Key pressed: 0x001d => 0xffe3
 Viewport:    Key pressed: 0x002f => 0x0076
 Viewport:    Key released: 0x002f => 0x0076
 Viewport:    Key released: 0x001d => 0xffe3

Thu Oct 28 11:37:17 2021
 Viewport:    Key pressed: 0x001d => 0xffe3
 Viewport:    Key pressed: 0x002f => 0x0076
 Viewport:    Key released: 0x002f => 0x0076
 Viewport:    Key released: 0x001d => 0xffe3

Thu Oct 28 11:37:18 2021
 CConn:       End of stream
```

**Comment by CendioOssman** [comment]
I'm afraid there is some network issue that is killing your connection:

> ```
> Oct 28 09:37:18 [servername] Xvnc[2097314]: VNCSConnST: closing Y.Y.Y.Y::62673: read: Connection reset by peer (104)
> ```

This error means that something abruptly killed the connection. I'm afraid we don't get any more details than that. It might be a local firewall or antivirus on the client or server. Or it might be a firewall on your network. I'm afraid you'll have to examine things between your server and client and see what might be causing the issue.

Sorry we can't provide any more help than that.

**Comment by masi-d** [comment]
I must say that I can not completely rule out that there may be a problem in the network, as the network between server and client is not under my personal control.
But for me it seems to be a software issue, as the problem arised with tigervnc-server version 1.11.0 and vanished after downgrade to 1.10.1.
Perhaps there is a timing issue which is handled more generously by version 1.10.1 than by version 1.11.0?

**Comment by CendioOssman** [comment]
I'm afraid that error code is not something that can be generated by the application itself. The difference between versions might be a firewall that tries to detect certain patterns, and that the old versions didn't trigger those.

**Comment by masi-d** [comment]
Indeed one of our firewalls was in charge for killing the connections.
Just for your information, the firewall claims that there is a buffer overrun attac detected in the vnc protocol.
I do not know if there is an error in tigervnc's implementation of the vnc protocol or in the detection pattern of the firewall.
I will investigate this further with our firewall people, to find out on which side the error is.

**Comment by CendioOssman** [comment]
Interesting. If you can find out more then we would definitely be interested. Do you know which make and model it is?

**Comment by masi-d** [comment]
The firewalls are from check point (https://www.checkpoint.com).
The exact model is unknown to me, if you really need to know which model it is, I can ask our firewall people.

The triggered filter is named "UltraVNC VNCViewer Authenticate Buffer Overflow".
More information is available at the manufacturers website:
https://threatpoint.checkpoint.com/ThreatPortal/threat?threatType=protection&threatId=AMSN20090203_17

**Comment by CendioOssman** [comment]
I'm not aware of any scenario where we send integers that cause overflows. So I think that product is misdetecting things.

Have you filed a support request with them about this? They'll probably listen more to a customer request than if we send them something.

**Comment by masi-d** [comment]
I will ask our firewall people if they can make a support request.

**Comment by masi-d** [comment]
I asked our firewall people to make a support request.
Unfortunately they are not willing to do this.
Their standing point is, that false positives happen from time to time and they have not the manpower to investigate this with the manufacturer.
For now they exclude my vnc server from this filter, to prevent the connection dropping in the future.
I will let them recheck if the filter triggers also when we have upgraded to TigerVNC 1.12.0

**Comment by CendioOssman** [comment]
Alright. Thanks for trying!

I've sent a request on my own to CheckPoint, by haven't heard anything back. We'll see if they are willing to discuss this.

**Comment by alanb32** [comment]
I have exactly the same issue, windows RDP closes after mouse click or sometime keyboard press - at random times, 
I am using radhat 7 with xrdp and tiger vnc server
I am very new to linux and want to know how I can downgrade the tiger vnc version?
can anyone help?

**Comment by masi-d** [comment]
As this seems to be an firewall issue, you should ask your firewall people to help.
The command to downgrade in redhat would be like this: `yum install tigervnc-server-1.10.1`

**Comment by alanb32** [comment]
Hi Again, 

I tried your downgrade of tigervnc
sudo yum install tigervnc-server-1.10.1
and rebooted and still does the same - its seems to close on right mouse click a lot, and sometimes on keyboard press

could there be a problem with xrdp or sesman ?

thanks for any help

**Comment by tiekookeit** [comment]
The same happens to me after upgrading to debian package version 1.11.0. I nmaped the vnc port and it was closed after get disconnected, but de processe was still running and no usefull message was logged to the log file. Sometimes it just get back to work after one or two minutes, but usually I have to kill the process. And that hapens once and while, not every time. But to me vnc just closes itself from serving requests.
